flink-connectors/flink-connector-datagen/src/test/java/org/apache/flink/connector/datagen/source/DataGeneratorSourceTest.java

```diff
@@ -99,23 +99,28 @@ void testReaderCheckpoints() throws Exception {
                         new NumberSequenceSource.NumberSequenceSplit("split-1", from, mid),
                         new NumberSequenceSource.NumberSequenceSplit("split-2", mid + 1, to)));
 
-        long remainingInCycle = elementsPerCycle;
-        while (reader.pollNext(out) != InputStatus.END_OF_INPUT) {
-            if (--remainingInCycle <= 0) {
-                remainingInCycle = elementsPerCycle;
-                // checkpoint
-                List<NumberSequenceSource.NumberSequenceSplit> splits = reader.snapshotState(1L);
-
-                // re-create and restore
-                reader = createReader();
-                if (splits.isEmpty()) {
-                    reader.notifyNoMoreSplits();
-                } else {
-                    reader.addSplits(splits);
-                }
+        for (int cycle = 0; cycle < 3; cycle++) {
+            // this call is not required but mimics what happens at runtime
+            reader.pollNext(out);
+            for (int elementInCycle = 0; elementInCycle < elementsPerCycle; elementInCycle++) {
+                reader.isAvailable().get();
+                reader.pollNext(out);
+            }
+            // checkpoint
+            List<NumberSequenceSource.NumberSequenceSplit> splits = reader.snapshotState(1L);
+
+            // re-create and restore
+            reader = createReader();
+            if (splits.isEmpty()) {
+                reader.notifyNoMoreSplits();
+            } else {
+                reader.addSplits(splits);
             }
         }
 
+        reader.isAvailable().get();
```

Comment:
```suggestion
```
Is this really a requirement? Looking at [StreamTask#processInput:561](https://github.com/apache/flink/blob/5601da7cc251eac479fe24167c0f58dbd963072f/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java#L561) shows that we're not looping over `isAvailable` anymore if reaching the end of data. Also from a logical standpoint, I would expect `isAvailable` to complete if we reach the end of the data stream. :thinking:

Reply:
yes and no.

> we're not looping over isAvailable anymore if reaching the end of data

Indeed, but the devil is in the details. Whether we reached the end of data isn't based on the input being exhausted, but on the source returning `END_OF_INPUT` on a call to pollNext().
And IteratorSourceReaderBase#pollNext (which the data generator source uses internally) doesn't return end-of-input when the input is exhausted, but when _another_ call was made to pollNext while the input was _already_ exhausted. (Which is allowed by the source reader interface)

That's why we need this additional call; the last call to pollNext within the loop does not `END_OF_INPUT` (as you expected I suppose) but `NOTHING_AVAILABLE`, and so we have to call it again.

It's all a bit finicky because the source reader api isn't intuitive in the slightest.

(fixing this in the IteratorSourceReaderBase isn't trivial :/)

Reply:
ah true, it relies on [IteratorSourceReaderBase#noMoreSplits](https://github.com/apache/flink/blob/a5667e82e25cb87dc5523b82b08aec3e1408e9c6/flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReaderBase.java#L136). Thanks for clarification. :+1: